src/main/java/org/jboss/arquillian/testcontainers/TestContainerProvider.java

```diff
@@ -16,15 +16,31 @@
 public class TestContainerProvider extends OperatesOnDeploymentAwareProvider {
     @Inject
     @ClassScoped
-    private Instance<GenericContainer<?>> genericContainerInstance;
+    private Instance<TestContainerInstances> containerInstances;
 
     @Override
     public Object doLookup(ArquillianResource resource, Annotation... qualifiers) {
-        return genericContainerInstance.get();
+        TestContainerInstances instances = containerInstances.get();
+        if (instances != null) {
+            // if there is only 1 container, return the instance
+            if (instances.all().size() == 1) {
+                return instances.get(0);
+            }
+            // if there is more then 1 container, search if there is a matching qualifier
+            for (GenericContainer<?> container : instances.all()) {
+                for (Annotation qualifier : qualifiers) {
+                    if (container.getClass().isAnnotationPresent(qualifier.annotationType())) {
+                        return container;
+                    }
+                }
+            }
+        }
+        // if there are more than 1 containers and no qualifier matches, return all instances
+        return instances;
```

Comment:
I think we need some more validation here. For example, we need to return `null` or throw an exception if there is an attempt to inject a `GenericContainer` into a `TestContainerInstances` or vice-versa.

Another option I've not looked to see how complicated it would be, would be to have `TestContainerInstances` implement a `GenericContainer`.

Reply:
I'll work on that, I already noticed in my test code that you can easily make it wrong, and the exception you get is not nice. My main problem here is how I can find out what the injection point (or its class) actually is - any idea how I can retrieve this with the given arguments in the doLookup? 

Should I store the type from the call in canProvide() and assume the doLookup() refers to it?

Reply:
FWIW I don't think we need to block the PR on this. We can file a follow up PR to check this. What we might actually end up needing is two different `ResourceProvider`'s. Which is likely not that big of a deal really.